diff -r ad2ee74b913a libpurple/protocols/facebook/Makefile.am
--- a/libpurple/protocols/facebook/Makefile.am	Thu Jan 07 14:06:04 2016 -0500
+++ b/libpurple/protocols/facebook/Makefile.am	Thu Jan 07 15:45:09 2016 -0500
@@ -20,7 +20,12 @@
 	thrift.c \
 	thrift.h \
 	util.c \
-	util.h
+	util.h \
+	../../glibcompat.h \
+	../../http.c \
+	../../http.h \
+	../../purple-socket.h \
+	../../purple-socket.c
 
 AM_CFLAGS = $(st)
 
@@ -43,10 +48,9 @@
 endif
 
 AM_CPPFLAGS = \
-	-I$(top_srcdir)/libpurple \
-	-I$(top_builddir)/libpurple \
-	-I$(top_srcdir) \
 	$(GLIB_CFLAGS) \
 	$(JSON_CFLAGS) \
-	$(GPLUGIN_CFLAGS) \
+	$(PURPLE_CFLAGS) \
+	$(ZLIB_CFLAGS) \
+	$(PLUGIN_CFLAGS) \
 	$(DEBUG_CFLAGS)
diff -r ad2ee74b913a libpurple/protocols/facebook/Makefile.mingw
--- a/libpurple/protocols/facebook/Makefile.mingw	Thu Jan 07 14:06:04 2016 -0500
+++ b/libpurple/protocols/facebook/Makefile.mingw	Thu Jan 07 15:45:09 2016 -0500
@@ -49,7 +49,9 @@
 			json.c \
 			mqtt.c \
 			thrift.c \
-			util.c
+			util.c \
+			../../http.c \
+			../../purple-socket.c
 
 OBJECTS = $(C_SRC:%.c=%.o)
 
@@ -62,7 +64,6 @@
 			-lgobject-2.0 \
 			-lws2_32 \
 			-lintl \
-			-lgplugin \
 			-ljson-glib-1.0 \
 			-lz \
 			-lpurple
@@ -76,12 +77,12 @@
 
 all: $(TARGET).dll
 
-install: all $(DLL_INSTALL_DIR)
+install: all
 	cp $(TARGET).dll $(DLL_INSTALL_DIR)
 
 $(OBJECTS): $(PURPLE_CONFIG_H)
 
-$(TARGET).dll: $(PURPLE_DLL).a $(OBJECTS)
+$(TARGET).dll: $(OBJECTS)
 	$(CC) -shared $(OBJECTS) $(LIB_PATHS) $(LIBS) $(DLL_LD_FLAGS) -o $(TARGET).dll
 
 ##
