diff -r 078fbf8d1a4e libpurple/protocols/facebook/Makefile.am
--- a/libpurple/protocols/facebook/Makefile.am	Mon Dec 21 22:29:34 2015 -0500
+++ b/libpurple/protocols/facebook/Makefile.am	Mon Dec 21 22:37:31 2015 -0500
@@ -23,7 +23,12 @@
 	thrift.c \
 	thrift.h \
 	util.c \
-	util.h
+	util.h \
+	../../glibcompat.h \
+	../../http.c \
+	../../http.h \
+	../../purple-socket.h \
+	../../purple-socket.c
 
 CLEANFILES = \
 	marshal.c \
@@ -57,10 +62,9 @@
 endif
 
 AM_CPPFLAGS = \
-	-I$(top_srcdir)/libpurple \
-	-I$(top_builddir)/libpurple \
 	$(GLIB_CFLAGS) \
 	$(JSON_CFLAGS) \
+	$(PURPLE_CFLAGS) \
 	$(ZLIB_CFLAGS) \
-	$(GPLUGIN_CFLAGS) \
+	$(PLUGIN_CFLAGS) \
 	$(DEBUG_CFLAGS)
diff -r 078fbf8d1a4e libpurple/protocols/facebook/Makefile.mingw
--- a/libpurple/protocols/facebook/Makefile.mingw	Mon Dec 21 22:29:34 2015 -0500
+++ b/libpurple/protocols/facebook/Makefile.mingw	Mon Dec 21 22:37:31 2015 -0500
@@ -50,7 +50,9 @@
 			json.c \
 			mqtt.c \
 			thrift.c \
-			util.c
+			util.c \
+			../../http.c \
+			../../purple-socket.c
 
 OBJECTS = $(C_SRC:%.c=%.o)
 
@@ -63,7 +65,6 @@
 			-lgobject-2.0 \
 			-lws2_32 \
 			-lintl \
-			-lgplugin \
 			-ljson-glib-1.0 \
 			-lz \
 			-lpurple
@@ -77,12 +78,12 @@
 
 all: $(TARGET).dll
 
-install: all $(DLL_INSTALL_DIR)
+install: all
 	cp $(TARGET).dll $(DLL_INSTALL_DIR)
 
 $(OBJECTS): $(PURPLE_CONFIG_H)
 
-$(TARGET).dll: $(PURPLE_DLL).a $(OBJECTS)
+$(TARGET).dll: $(OBJECTS)
 	$(CC) -shared $(OBJECTS) $(LIB_PATHS) $(LIBS) $(DLL_LD_FLAGS) -o $(TARGET).dll
 
 marshal.c: marshaller.list marshal.h
