diff --git a/libpurple/http.c b/libpurple/http.c
--- a/libpurple/http.c
+++ b/libpurple/http.c
@@ -546,11 +546,7 @@
 	cb_data = g_object_steal_data(source, "cb_data");
 
 	if (conn == NULL) {
-		if (!g_error_matches(error,
-				G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
-			cb(hs, error->message, cb_data);
-		}
-
+		cb(hs, error->message, cb_data);
 		g_clear_error(&error);
 		return;
 	}
@@ -1223,10 +1219,8 @@
 				&error);
 	got_anything = (len > 0);
 
-	if (len < 0 && (g_error_matches(error,
-			G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK) ||
-			g_error_matches(error,
-			G_IO_ERROR, G_IO_ERROR_CANCELLED))) {
+	if (len < 0 && g_error_matches(error,
+			G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK)) {
 		g_clear_error(&error);
 		return FALSE;
 	}
@@ -1511,10 +1505,8 @@
 				&error);
 	}
 
-	if (written < 0 && (g_error_matches(error,
-			G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK) ||
-			g_error_matches(error,
-			G_IO_ERROR, G_IO_ERROR_CANCELLED))) {
+	if (written < 0 && g_error_matches(error,
+			G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK)) {
 		g_clear_error(&error);
 		return G_SOURCE_CONTINUE;
 	}
