--- a/libpurple/protocols/facebook/api.h	2016-11-21 09:56:02.734212775 -0300
+++ b/libpurple/protocols/facebook/api.h	2016-11-21 10:14:38.808696834 -0300
@@ -91,6 +91,13 @@
 #define FB_API_SECRET  "374e60f8b9bb6b8cbb30f78030438895"
 
 /**
+ * FB_API_AGENT:
+ *
+ * The HTTP User-Agent header.
+ */
+#define FB_API_AGENT  "Facebook plugin / Purple / 0.9"
+
+/**
  * FB_API_URL_ATTACH:
  *
  * The URL for attachment URL requests.
--- a/libpurple/protocols/facebook/api.c	2016-11-21 09:56:02.734212775 -0300
+++ b/libpurple/protocols/facebook/api.c	2016-11-21 10:15:05.214906611 -0300
@@ -762,6 +762,8 @@
 		g_free(data);
 	}
 
+	purple_http_request_header_set(req, "User-Agent", FB_API_AGENT);
+
 	data = fb_http_params_close(params, NULL);
 	purple_http_request_set_contents(req, data, -1);
 	ret = purple_http_request(priv->gc, req, callback, api);
